src/builtins/history.cpp

```diff
@@ -260,6 +260,15 @@ maybe_t<int> builtin_history(parser_t &parser, io_streams_t &streams, const wcha
                                  parser.cancel_checker(), streams)) {
                 status = STATUS_CMD_ERROR;
             }
+            bool no_results = true; // TODO how can we find this?
```

Comment:
I don't know how to check whether there were no results. `history_t::search` seems to write directly to an IO stream, and returns `false` only when the user supplies invalid arguments.

Reply:
it's easier if we move this logic further down the call stack:

```diff
diff --git a/src/history.cpp b/src/history.cpp
index c07a7ebb9..e8a7d7913 100644
--- a/src/history.cpp
+++ b/src/history.cpp
@@ -1475,9 +1475,16 @@ static void do_1_history_search(history_t *hist, history_search_type_t search_ty
     history_search_t searcher = history_search_t(hist, search_string, search_type,
                                                  case_sensitive ? 0 : history_search_ignore_case);
+    bool success = false;
     while (!cancel_check() && searcher.go_to_next_match(history_search_direction_t::backward)) {
+        success = true;
         if (!func(searcher.current_item())) {
             break;
         }
     }
+    if (!success && (search_type == history_search_type_t::contains ||
+                     search_type == history_search_type_t::contains_glob)) {
+        do_1_history_search(hist, history_search_type_t::contains_subsequence, search_string,
+                            case_sensitive, func, cancel_check);
+    }
 }
 
```